### ISSUE TITLE
Build Time Optimization 

### ISSUE THREAD

**Issue by cbielow** [OpenMS/OpenMS#4341]
## via #include reduction
we should look into optimizing our build time by removing unneccessary includes, especially from widely used headers (as partially done in #4333 and #4340).

Good candidates are all boost includes (especially when used in our headers), our own LogStream.h and probably a gazillion others.
We should make use of forward declarations when possible, simply by `class <z>` or for more involved things e.g. `<boost/regex_fwd.hpp>` and `<iosfwd>` (note that STL forward declarations are highly discouraged/forbidden), and maybe use PIMPL idioms a bit more (mind performance).

These are some good starting points:
https://arne-mertz.de/2018/03/forward-declaring-templates-and-enums/
https://blog.magnum.graphics/backstage/lightweight-stl-compatible-unique-pointer/

## via reduction of number of warnings
https://stackoverflow.com/questions/25382568/can-a-large-number-of-warnings-increase-compilation-time

preliminary results (incomplete):
the overhead of printing the warnings is not much:
Enabling `-weverything` in Clang (which prints 952 Warnings when building OpenMS lib and produces 606 KB of output vs. 81KB output without warnings) increases build time from 41m:34s to 41m:37s (i.e. 3 seconds). Actually fixing the warnings will probably not help much more.
 todo: test the same on VS (which is known to be slow with console I/O).

## via reducing the parsing of useless code
e.g. removing the copyright header (33 lines each) in every .hpp/.cpp file and replacing it with something slim

**Comment by hroest** [comment]
In general I think this is a great idea, especially when its boost and other rarely used headers. We could start with PIMPL for some classes where attributes are optional and space for an empty pointer (8 bytes) uses less space than an empty vector/map. So we would gain space as well

**Comment by timosachsenberg** [comment]
Potentially also interesting read: http://oliora.github.io/2015/12/29/pimpl-and-rule-of-zero.html
How to get rid of most boilerplate code using PIMPL.

**Comment by timosachsenberg** [comment]
or do some general benchmarking of build times (there are some options out there)

**Comment by VyasGuru** [comment]
hi, is this issue still open to pursue?

**Comment by timosachsenberg** [comment]
Yes. Still an issue ;)

**Comment by VyasGuru** [comment]
hi, could you suggest a chunk of the code to test this out on? just want to test it out before approaching doing it on a wider scale. thanks

**Comment by timosachsenberg** [comment]
I would suggest you look into build time profiling for different compiler and then build openms. It should spit out a file that tells you where the compiler spend most time.
see e.g. https://github.com/aras-p/ClangBuildAnalyzer for additional information

**Comment by VyasGuru** [comment]
idk if this is a mac only issue but the version of coinor that it tries to install during build doesnt work because it uses an http site, could work around it but is it kept that outdated site for some reason?

**Comment by timosachsenberg** [comment]
I think it works in our ci / GitHub actions. Maybe check in the ci script how we install it on Mac.

**Comment by codemaster1104** [comment]
Hey, this seems interesting , was deep diving with clang will update my findings soon

**Comment by codemaster1104** [comment]
btw this is the current build report [doc](https://docs.google.com/document/d/1xAdAuwgmomoCNV_HE_l0jYTLT9Wu1R5SDqlcMccrVw0/edit?usp=sharing)

**Comment by timosachsenberg** [comment]
Do you see any easy to implement changes that could improve them?

**Comment by codemaster1104** [comment]
One of the quick optimization would be to pre compile expensive headers 
`835009 ms: /usr/include/c++/11/string  (1665 times, avg 501 ms)
642735 ms: OpenMS/DATASTRUCTURES/String.h  (1566 times, avg 410 ms)
659962 ms: PrecisionWrapper.h  (1348 times, avg 489 ms)`

So i added safe headers like string, vectors to precompiled headers in cmake

`target_precompile_headers(OpenMS PRIVATE
  <string>
  <vector>
  <map>
  <unordered_map>
  <OpenMS/DATASTRUCTURES/String.h>
  <OpenMS/CONCEPT/PrecisionWrapper.h>
)`

It did improve the parse time in build 
Should I make a PR on this so you can test it too?

```
Metric	Before	After	Change
Frontend parse time	10,237s	10,046s	−191s (−1.9%)
```
*my pc was throttling during the build so result may actually be better

**Comment by timosachsenberg** [comment]
if we don't cache those I guess we will not have much wins in our CI?
Not sure if worth the effort then.

**Comment by cbielow** [comment]
You could give forward declarations in FileHandler.h a try. Its a very central header and it could greatly benefit from forward declarations. You will need to add more includes in other places, after you have removed the includes in FileHandler.h, but in total it should be a win, since includes are more targeteted in places where they are actually needed.

**Comment by codemaster1104** [comment]
> if we don't cache those I guess we will not have much wins in our CI? Not sure if worth the effort then.

yeah on CI it wont be beneficial but on local machines for incremental builds it will improve the build times.
I had already implemented it so will just raise the pr if its fine as in any way its not downgrading the performance in CI while improving local machine performance

**Comment by codemaster1104** [comment]
> You could give forward declarations in FileHandler.h a try. Its a very central header and it could greatly benefit from forward declarations. You will need to add more includes in other places, after you have removed the includes in FileHandler.h, but in total it should be a win, since includes are more targeteted in places where they are actually needed.

ok will look into it next

**Comment by timosachsenberg** [comment]
@jpfeuffer does it make sense to use the precompiled headers? assuming people that care about build times use ccache?

**Comment by codemaster1104** [comment]
> You could give forward declarations in FileHandler.h a try. Its a very central header and it could greatly benefit from forward declarations. You will need to add more includes in other places, after you have removed the includes in FileHandler.h, but in total it should be a win, since includes are more targeteted in places where they are actually needed.

Implemented this
basically FileHandler.h is included by 130+ translation units and was
transitively pulling in several heavy headers:

  - ConsensusMap.h
  - ProteinIdentification.h
  - PeptideIdentification.h
  - PeptideIdentificationList.h

These types were only used as references in method signatures and
therefore did not require full definitions in the header.

Replaced the includes with forward declarations and removed the
ConsensusMap() default argument from storeQC(), which was the only
case requiring a complete type in the header.

~100 downstream .cpp files that relied on transitive includes were
updated to include the required headers explicitly (automated using script).

Impact (ClangBuildAnalyzer):
  Parsing (frontend):  9903.4 s
  Codegen (backend):   3411.3 s

**Comment by jpfeuffer** [comment]
> [@jpfeuffer](https://github.com/jpfeuffer) does it make sense to use the precompiled headers? assuming people that care about build times use ccache?

I would say if we can be sure that it does not affect anything else and it is really just those two cmake lines, it doesn't hurt.
But I am not sure at which point we can be sure 😃

**Comment by codemaster1104** [comment]
Hey, next i am planning to give forward declarations in ConsensusMap.h a try considering the clang build report
